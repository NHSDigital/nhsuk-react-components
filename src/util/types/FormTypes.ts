import { HTMLProps } from 'react';
import { ErrorMessageProps } from '@components/form-elements/error-message/ErrorMessage';
import { HintTextProps } from '@components/form-elements/hint-text/HintText';
import { LabelProps } from '@components/form-elements/label/Label';

export interface FormElementProps {
  label?: string;
  labelProps?: LabelProps;
  error?: string | boolean;
  errorProps?: ErrorMessageProps;
  hint?: string;
  hintProps?: HintTextProps;
  formGroupProps?: HTMLProps<HTMLDivElement>;
  disableErrorLine?: boolean;
  id?: string;
  name?: string;
}
