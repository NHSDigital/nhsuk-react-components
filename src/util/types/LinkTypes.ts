import { AnchorHTMLAttributes, ButtonHTMLAttributes, ElementType } from 'react';

export type AsElementLink<T extends HTMLElement> = (T extends HTMLButtonElement
  ? ButtonHTMLAttributes<T>
  : AnchorHTMLAttributes<T>) & {
  asElement?: ElementType;
  to?: string;
};
