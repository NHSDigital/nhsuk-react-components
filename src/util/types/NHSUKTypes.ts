export type NHSUKSize = 's' | 'm' | 'l' | 'xl';

export type InputWidth = '2' | '3' | '4' | '5' | '10' | '20' | '30' | 2 | 3 | 4 | 5 | 10 | 20 | 30;

export type CareCardType = 'non-urgent' | 'urgent' | 'emergency';

export type CardType = 'feature' | 'primary' | 'secondary' | CareCardType;

export type ColWidth =
  | 'full'
  | 'three-quarters'
  | 'one-half'
  | 'two-thirds'
  | 'one-third'
  | 'one-quarter';

export type HTMLAttributesWithData<T> = React.HTMLAttributes<T> & {
  [key: `data-${string}`]: unknown;
};
