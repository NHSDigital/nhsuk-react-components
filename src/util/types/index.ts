export * from './FormTypes.js';
export * from './LinkTypes.js';
export * from './NHSUKTypes.js';
export * from './TypeGuards.js';
