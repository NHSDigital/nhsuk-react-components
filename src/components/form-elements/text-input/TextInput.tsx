import React, { ComponentPropsWithoutRef, FC, forwardRef } from 'react';
import classNames from 'classnames';
import FormGroup from '@components/utils/FormGroup';
import { InputWidth } from '@util/types/NHSUKTypes';
import { FormElementProps } from '@util/types/FormTypes';

export interface TextInputProps
  extends ComponentPropsWithoutRef<'input'>,
    Omit<FormElementProps, 'fieldsetProps' | 'legend' | 'legendProps'> {
  width?: InputWidth;
  prefix?: string;
  suffix?: string;
}

const TextInputPrefix: FC<Pick<TextInputProps, 'prefix'>> = ({ prefix }) => (
  <div className="nhsuk-input__prefix" aria-hidden="true">
    {prefix}
  </div>
);

const TextInputSuffix: FC<Pick<TextInputProps, 'suffix'>> = ({ suffix }) => (
  <div className="nhsuk-input__suffix" aria-hidden="true">
    {suffix}
  </div>
);

const TextInputComponent = forwardRef<HTMLInputElement, TextInputProps>((props, forwardedRef) => (
  <FormGroup<TextInputProps> {...props} inputType="input">
    {({ width, className, error, type = 'text', prefix, suffix, ...rest }) => {
      const Input = (
        <input
          className={classNames(
            'nhsuk-input',
            { [`nhsuk-input--width-${width}`]: width },
            { 'nhsuk-input--error': error },
            className,
          )}
          ref={forwardedRef}
          type={type}
          {...rest}
        />
      );

      return prefix || suffix ? (
        <div className="nhsuk-input__wrapper">
          {prefix ? <TextInputPrefix prefix={prefix} /> : null}
          {Input}
          {suffix ? <TextInputSuffix suffix={suffix} /> : null}
        </div>
      ) : (
        Input
      );
    }}
  </FormGroup>
));

TextInputComponent.displayName = 'TextInput';

export default TextInputComponent;
