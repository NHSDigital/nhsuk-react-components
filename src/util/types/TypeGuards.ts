/* eslint-disable @typescript-eslint/no-explicit-any */
import { FC, ReactElement, JSXElementConstructor, ReactNode, ReactPortal } from 'react';
import { CardType, CareCardType } from './NHSUKTypes';

/**
 * Assert that a child item is of the given component type.
 */
export const childIsOfComponentType = (
  child:
    | string
    | number
    | boolean
    | ReactElement<any, string | JSXElementConstructor<any>>
    | Iterable<ReactNode>
    | ReactPortal
    | null
    | undefined,
  component: FC,
): child is
  | React.ReactElement<any, string | React.JSXElementConstructor<any>>
  | React.ReactPortal =>
  child !== undefined &&
  child !== null &&
  typeof child === 'object' &&
  'type' in child &&
  child.type === component;

/**
 * Check whether the given card type is that of a care card.
 */
export const cardTypeIsCareCard = (cardType: CardType | undefined): cardType is CareCardType =>
  cardType === 'non-urgent' || cardType === 'urgent' || cardType === 'emergency';
