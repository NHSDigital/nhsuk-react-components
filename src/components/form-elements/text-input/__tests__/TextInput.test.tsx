import React, { useRef } from 'react';
import { fireEvent, render } from '@testing-library/react';
import TextInput from '../TextInput';
import { InputWidth } from '@util/types/NHSUKTypes';

describe('TextInput', () => {
  afterEach(() => {
    jest.restoreAllMocks();
  });

  const TextInputComp = ({ onHandle }: { onHandle: () => void }) => {
    const ref = useRef(null);
    const handleClick = () => {
      if (!ref.current) return;
      onHandle();
    };

    return <TextInput type="button" onClick={handleClick} inputRef={ref} />;
  };

  it('should handle click where ref Exists', () => {
    const useRefSpy = jest
      .spyOn(React, 'useRef')
      .mockReturnValueOnce({ current: document.createElement('button') });
    const mock = jest.fn();
    const { container } = render(<TextInputComp onHandle={mock} />);

    const textInputEl = container.querySelector('input')!;
    fireEvent.click(textInputEl);

    expect(useRefSpy).toBeCalledWith(null);
    expect(mock).toBeCalledTimes(1);
  });

  it.each<InputWidth | undefined>([undefined, '5', '10'])(
    'Sets the provided input width if specified with %s',
    (width) => {
      const { container } = render(<TextInput width={width} />);

      const input = container.querySelector('.nhsuk-input');

      if (width) {
        expect(input).toHaveClass(`nhsuk-input--width-${width}`);
      } else {
        expect(input?.className.indexOf('nhsuk-input--width')).toBe(-1);
      }
    },
  );

  it.each<string | boolean | undefined>([undefined, true, false, 'error'])(
    'Sets the error class if specified with %s',
    (error) => {
      const { container } = render(<TextInput error={error} />);

      const input = container.querySelector('.nhsuk-input');

      if (error) {
        expect(input).toHaveClass('nhsuk-input--error');
      } else {
        expect(input).not.toHaveClass('nhsuk-input--error');
      }
    },
  );

  it('Sets the provided input width if specified', () => {
    const { container } = render(<TextInput width="5" />);

    expect(container.querySelector('.nhsuk-input')).toHaveClass('nhsuk-input--width-5');
  });

  it.each`
    prefix       | suffix
    ${'£'}       | ${'per item'}
    ${'£'}       | ${undefined}
    ${undefined} | ${'per item'}
    ${undefined} | ${undefined}
  `(
    'Renders expected prefix and suffix when provided $prefix and $suffix',
    ({ prefix, suffix }) => {
      const { container } = render(<TextInput prefix={prefix} suffix={suffix} />);

      const prefixElement = container.querySelector('.nhsuk-input__wrapper > .nhsuk-input__prefix');
      const suffixElement = container.querySelector('.nhsuk-input__wrapper > .nhsuk-input__suffix');

      if (prefix) {
        expect(prefixElement).not.toBeNull();
        expect(prefixElement?.textContent).toBe(prefix);
      } else {
        expect(prefixElement).toBeNull();
      }

      if (suffix) {
        expect(suffixElement).not.toBeNull();
        expect(suffixElement?.textContent).toBe(suffix);
      } else {
        expect(suffixElement).toBeNull();
      }

      if (!prefix && !suffix) {
        expect(container.querySelector('.nhsuk-input__wrapper')).toBeNull();
        expect(container.querySelector('.nhsuk-input')).not.toBeNull();
      } else {
        expect(container.querySelector('.nhsuk-input__wrapper > .nhsuk-input')).not.toBeNull();
      }
    },
  );
});
