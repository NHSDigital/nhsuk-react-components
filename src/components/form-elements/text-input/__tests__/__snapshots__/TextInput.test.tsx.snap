// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextInput matches snapshot: TextInput 1`] = `
<div>
  <div
    class="nhsuk-form-group"
  >
    <label
      class="nhsuk-label nhsuk-label--l"
      for="nhs-number"
      id="nhs-number--label"
    >
      What is your NHS number?
    </label>
    <input
      class="nhsuk-input"
      id="nhs-number"
      name="nhs-number"
      type="text"
    />
  </div>
</div>
`;
